comRamona/assignment2/Question2.java

```diff
@@ -0,0 +1,108 @@
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Design an algorithm and write ​ code​ to find the lowest common ancestor of two nodes in a binary tree.
+ * Avoid storing additional nodes in a data structure.
+ */
+public class Question2 {
+    static boolean foundOne;
+    static boolean foundTwo;
+
+    /**
+     * Helper method to find lowest common ancestor of two nodes
+     *
+     * @param currentNode
+     * @param nodeOne
+     * @param nodeTwo
+     * @return lowest common ancestor or one of the nodes if not both present
+     */
+    private static MyNode findLowestCommonAncestorHelper(MyNode currentNode, MyNode nodeOne, MyNode nodeTwo) {
+        if (currentNode == null) {
+            return null;
+        }
+        if (currentNode.equals(nodeOne)) {
+            foundOne = true;
+            // could stop search and return here, this is just to make sure both nodes are in the tree
+            if (!foundTwo) {
+                findLowestCommonAncestorHelper(currentNode.getLeft(), nodeOne, nodeTwo);
+                findLowestCommonAncestorHelper(currentNode.getRight(), nodeOne, nodeTwo);
+            }
+            return currentNode;
+        }
+        if (currentNode.equals(nodeTwo)) {
+            foundTwo = true;
+            // could stop search and return here, this is just to make sure both nodes are in the tree
+            if (!foundOne) {
+                findLowestCommonAncestorHelper(currentNode.getLeft(), nodeOne, nodeTwo);
+                findLowestCommonAncestorHelper(currentNode.getRight(), nodeOne, nodeTwo);
+            }
+            return currentNode;
+        }
+
+        MyNode leftSearch = findLowestCommonAncestorHelper(currentNode.getLeft(), nodeOne, nodeTwo);
+        MyNode rightSearch = findLowestCommonAncestorHelper(currentNode.getRight(), nodeOne, nodeTwo);
+        if (leftSearch != null && rightSearch != null) {
+            return currentNode;
+        }
+        if (leftSearch != null) {
+            return leftSearch;
+        }
+        if (rightSearch != null) {
+            return rightSearch;
+        }
+        return null;
+    }
+
+    /**
+     * @param currentNode current root of search
+     * @param nodeOne     first node
+     * @param nodeTwo     second node
+     * @return common ancestor of first and second node or null if one node is not present
+     */
+    public static MyNode findLowestCommonAncestor(MyNode currentNode, MyNode nodeOne, MyNode nodeTwo) {
+        foundOne = false;
+        foundTwo = false;
+        MyNode result = findLowestCommonAncestorHelper(currentNode, nodeOne, nodeTwo);
```

Comment:
Can you think of how to use the findAncestors() answer to Part 1 to solve part 2 without writing the algorithm again.